File: src/main/java/org/example/kalkulationsprogramm/service/ZugferdErstellService.java
package org.example.kalkulationsprogramm.service;

import org.example.kalkulationsprogramm.dto.Zugferd.ZugferdDaten;
import org.mustangproject.Invoice;
import org.mustangproject.Item;
import org.mustangproject.Product;
import org.mustangproject.TradeParty;
import org.mustangproject.ZUGFeRD.ZUGFeRDExporterFromA3;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.LocalDate;
import java.time.ZoneId;
import java.util.Date;

@Service
public class ZugferdErstellService {

    private static Date toDate(LocalDate ld) {
        return ld == null ? null : Date.from(ld.atStartOfDay(ZoneId.systemDefault()).toInstant());
    }

    public Path erzeuge(String originalPdfPath, ZugferdDaten daten) {
        try {
            Path ziel = Files.createTempFile("zugferd-", ".pdf.html");

            try (ZUGFeRDExporterFromA3 exporter = new ZUGFeRDExporterFromA3()
                    // .setProfile(Profile.EN16931) // weggelassen → vermeidet Versionskonflikte
                    .setCreator("Kalkulationsprogramm")
                    .setProducer("Kalkulationsprogramm")
                    .load(originalPdfPath)) {

                Invoice invoice = new Invoice();
                String dokumentart = normalizeDokumentenart(daten.getGeschaeftsdokumentart());
                invoice.setNumber(daten.getRechnungsnummer());

                // ZUGFeRD requires an issue date - use today's date as fallback
                LocalDate issueDate = daten.getRechnungsdatum() != null
                        ? daten.getRechnungsdatum()
                        : LocalDate.now();
                invoice.setIssueDate(toDate(issueDate));
                if (istRechnung(dokumentart) && daten.getFaelligkeitsdatum() != null) {
                    invoice.setDueDate(toDate(daten.getFaelligkeitsdatum()));
                }

                TradeParty seller = new TradeParty();
                seller.setName("Example Company"); // Pflicht: Name

                TradeParty buyer = new TradeParty();
                buyer.setName(daten.getKundenName() != null ? daten.getKundenName() : "Kunde");
                if (daten.getKundennummer() != null) {
                    buyer.setID(daten.getKundennummer());
                }

                invoice.setSender(seller);
                invoice.setRecipient(buyer);

                BigDecimal betrag = daten.getBetrag() != null ? daten.getBetrag() : BigDecimal.ZERO;

                Product product = new Product();
                product.setName(dokumentart);
                product.setVATPercent(new BigDecimal("19")); // oder BigDecimal.ZERO bei steuerfrei
                // optional: product.setUnit("C62"); // Stück

                Item item = new Item(product, BigDecimal.ONE, betrag);
                invoice.addItem(item);

                invoice.addItem(item);

                exporter.setTransaction(invoice);
                exporter.export(ziel.toString());
            }

            return ziel;
        } catch (Exception e) {
            throw new RuntimeException("ZUGFeRD Erstellung fehlgeschlagen", e);
        }
    }

    private String normalizeDokumentenart(String art) {
        if (art == null) {
            return "Rechnung";
        }
        String trimmed = art.trim();
        return trimmed.isEmpty() ? "Rechnung" : trimmed;
    }

    private boolean istRechnung(String art) {
        return "rechnung".equalsIgnoreCase(art);
    }
}

File: src/main/kotlin/org/example/kalkulationsprogramm/service/ZugferdErstellService.kt
package org.example.kalkulationsprogramm.service

import org.example.kalkulationsprogramm.dto.Zugferd.ZugferdDaten
import org.mustangproject.Invoice
import org.mustangproject.Item
import org.mustangproject.Product
import org.mustangproject.TradeParty
import org.mustangproject.ZUGFeRD.ZUGFeRDExporterFromA3
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.nio.file.Files
import java.nio.file.Path
import java.time.LocalDate
import java.time.ZoneId
import java.util.Date

@Service
class ZugferdErstellService {

    fun erzeuge(originalPdfPath: String, daten: ZugferdDaten): Path {
        try {
            val ziel = Files.createTempFile("zugferd-", ".pdf.html")

            ZUGFeRDExporterFromA3()
                .setCreator("Kalkulationsprogramm")
                .setProducer("Kalkulationsprogramm")
                .load(originalPdfPath)
                .use { exporter ->
                    val invoice = Invoice()
                    val dokumentart = normalizeDokumentenart(daten.geschaeftsdokumentart)
                    invoice.setNumber(daten.rechnungsnummer)

                    val issueDate = daten.rechnungsdatum ?: LocalDate.now()
                    invoice.setIssueDate(toDate(issueDate))
                    if (istRechnung(dokumentart) && daten.faelligkeitsdatum != null) {
                        invoice.setDueDate(toDate(daten.faelligkeitsdatum))
                    }

                    val seller = TradeParty()
                    seller.setName("Bauschlosserei Kuhn")

                    val buyer = TradeParty()
                    buyer.setName(daten.kundenName ?: "Kunde")
                    daten.kundennummer?.let { buyer.setID(it) }

                    invoice.setSender(seller)
                    invoice.setRecipient(buyer)

                    val betrag = daten.betrag ?: BigDecimal.ZERO
                    val product = Product()
                    product.setName(dokumentart)
                    product.setVATPercent(BigDecimal("19"))

                    val item = Item(product, BigDecimal.ONE, betrag)
                    invoice.addItem(item)
                    invoice.addItem(item)

                    exporter.setTransaction(invoice)
                    exporter.export(ziel.toString())
                }

            return ziel
        } catch (e: Exception) {
            throw RuntimeException("ZUGFeRD Erstellung fehlgeschlagen", e)
        }
    }

    private fun normalizeDokumentenart(art: String?): String {
        if (art == null) {
            return "Rechnung"
        }
        val trimmed = art.trim()
        return if (trimmed.isEmpty()) "Rechnung" else trimmed
    }

    private fun istRechnung(art: String): Boolean =
        "rechnung".equals(art, ignoreCase = true)

    private companion object {
        fun toDate(ld: LocalDate?): Date? =
            ld?.let { Date.from(it.atStartOfDay(ZoneId.systemDefault()).toInstant()) }
    }
}
